test/unit/features/options/mixins.spec.js

```diff
@@ -0,0 +1,77 @@
+import Vue from 'vue'
+const mergeOptions = Vue.util.mergeOptions
+
+describe('Mixins', () => {
+  it('vm should have options from mixin', () => {
+    const mixin = {
+      directives: {
+        c: {}
+      },
+      methods: {
+        a: function () {}
+      }
+    }
+
+    const vm = new Vue({
+      mixins: [mixin],
+      methods: {
+        b: function () {}
+      }
+    }).$mount()
+
+    expect(vm.a).toBeDefined()
+    expect(vm.b).toBeDefined()
+    expect(vm.$options.directives.c).toBeDefined()
+  })
+
+  it('should call hooks from mixins first, mixin methods should not override defined method', () => {
```

Comment:
Can we split this into two separate cases? One for hook order and the other for method overriding.